### DISCUSSION TITLE
getting trial_duration into the data for a trial (a question from the RT task tutorial)

### DISCUSSION THREAD

**Discussion by CarolynBuck-Gengler** [jspsych/jsPsych#808] [Q&A] [answered]
I've been working my way through the RT tutorial after rewatching the workshop and doing the lexical decision task with some additional tweaks.

In Part 7, a function is added to generate different timings for the fixation cross, but that value is not part of the data generated for that trial. Is there a way to get that value into the data? Somehow I'm not able to get my head around how that value might be accessed. I'm assuming that if I could figure out how to access that value, I would put an on_finish callback function at the end of the definition of the fixation trial type.

thanks,
Carolyn

**Comment by jodeleeuw** [comment, marked as answer]
Hi Carolyn,

You're right -- this is not a straightforward value to access. I'm hoping to improve the data handling in the future so that any trial parameters can be optionally added onto the data object for a trial.

There are a couple workarounds. One is to assign this value to a global variable, and then add it in the `on_finish` callback.

```js
var fixation_timing = null; // create global variable
var trial = {
  type: 'html-keyboard-response',
  stimulus: '<div style="font-size:60px;">+</div>',
  choices: jsPsych.NO_KEYS,
  trial_duration: function(){
    // no var keyword means that it will assign to the global we created outside the trial definition
    fixation_timing = jsPsych.randomization.sampleWithoutReplacement([250, 500, 750, 1000, 1250, 1500, 1750, 2000], 1)[0];
    return fixation_timing;
  },
  on_finish: function(data){
    data.trial_duration = fixation_timing;
  }
}
```

Another option is to use the `on_start` callback ([docs here](https://www.jspsych.org/overview/callbacks/#on_start-trial)). This has the value of the parameter and you can modify the data object.

```js
var trial = {
  type: 'html-keyboard-response',
  stimulus: '<div style="font-size:60px;">+</div>',
  choices: jsPsych.NO_KEYS,
  trial_duration: function(){
    return jsPsych.randomization.sampleWithoutReplacement([250, 500, 750, 1000, 1250, 1500, 1750, 2000], 1)[0];
  },
  on_start: function(trial){
    trial.data.trial_duration = trial.trial_duration;
  }
}
```

**Reply by CarolynBuck-Gengler** [reply]
Thank you, I'll try both of those :-)

... Both work really well. The global variable version has the added bonus of being able to add the fixation timing to the actual trial of interest itself, at least in the simple case of the RT tutorial, which is even better.
Thanks again!
